api/authentication.py

```diff
@@ -0,0 +1,66 @@
+from datetime import datetime
+import json
+import requests
+
+from django.conf import settings
+from django.core.cache import cache
+
+from allauth.socialaccount.models import SocialAccount
+from rest_framework.authentication import (
+    BaseAuthentication, get_authorization_header
+)
+
+
+def get_cache_key(token):
+    return f'fxa_token_{token}'
+
+
+class FxaTokenAuthentication(BaseAuthentication):
+    def authenticate(self, request):
+        authorization = get_authorization_header(request).decode()
+        if not authorization or not authorization.startswith('Bearer '):
+            return None
+
+        token = request.headers.get('Authorization').split(' ')[1]
+        cache_key = get_cache_key(token)
+        fxa_resp = cache.get(cache_key)
+        if not fxa_resp:
+            introspect_token_url = (
+                '%s/introspect' %
+                settings.SOCIALACCOUNT_PROVIDERS['fxa']['OAUTH_ENDPOINT']
+            )
+            fxa_resp = requests.post(introspect_token_url, json={'token': token})
+            if not fxa_resp.status_code == 200:
+                # cache anything besides 200 for 60s - it might be an error
+                # we need to re-try, but we don't want to send un-throttled
+                # retries at FXA
+                cache.set(cache_key, fxa_resp, 60)
+                return None
+
+        fxa_resp_json = json.loads(fxa_resp.content)
```

Comment:
This can fail in lots of ways, but mainly if the response is HTML, such as a 5xx response from Apache / gunicorn. You should use [fxa_resp.json()](https://docs.python-requests.org/en/latest/user/quickstart/#json-response-content) and a try/except block.